### ISSUE TITLE
Adding shapes to a Body dynamically does not project shapes in isometric maps.

### ISSUE THREAD

**Issue by ldd** [melonjs/melonJS#601]
This applies for melonJS 1.2.0-beta

for reference, please look at 
https://groups.google.com/forum/#!topic/melonjs/Tz8tUDHgDGw

I wanted to extend Entity objects without using Tilled.

This works perfectly, as done in the above forum post, but for 1.2 I needed to do
settings = {
//bunch of settings
orientation: "isometric",
getTMXShapes: new me.TMXObject('empty').getTMXShapes;
}
this._super(....,'init', [x,y,settings]);

With this small workaround, everything works perfectly and flawlessly.

put in other words:
either orientation or/and getTMXShapes are required in the settings when creating an Entity.

**Comment by agmcleod** [comment]
We should probably create a new issue dictating what needs to be done. I know we've looked at a bit of the shapes & tmx object data lately. What should be done in order to satisfy isometric shapes?

**Comment by parasyte** [comment]
I thought it was pretty clear by the description. But to describe further, when a shape is added to an isometric map (say a rectangle) it needs to be projected by rotation and scaling: https://github.com/melonjs/melonJS/blob/0c173cbacbaf94441157c5701068352afb413087/src/level/TMXObjectGroup.js#L387-L392

An ideal solution would be to do the projection only when drawing. (Leave the shape and positioning internally to simplify entity movement.) The initial work (linked above) was done as a stopgap to ensure isometric map support wasn't completely missing. This ideal solution should be tracked in a different ticket. But completing the stopgap would happen here.

**Comment by agmcleod** [comment]
> An ideal solution would be to do the projection only when drawing. (Leave the shape and positioning internally to simplify entity movement.)

Makes sense. That way you're not dealing with "isometric units" directly. Though do physic shapes get saved as projected? Could be weird to have onething projected, and another note.

> This ideal solution should be tracked in a different ticket. But completing the stopgap would happen 
here.

alrighty

**Comment by parasyte** [comment]
In the ideal world, logical coordinates are always distinct from render coordinates; The axes do not necessarily point in any specific direction (relative to the screen) in a logical coordinate system. The shapes defined in the TMX are expressed in logical coordinates, which are roughly orthogonal. It's easiest to use these shapes directly for collision detection and entity movement. In other words, logical coordinates would be "native" to melonJS in that mode of operation.

That's the eventual goal, anyway, since it breaks the coupling between world physics and renderer; theoretically, a game can be rendered in 2D or 3D just by changing the drawing code, while everything else remains untouched.

**Comment by agmcleod** [comment]
Provided you put in 3d geometry ;)

**Comment by parasyte** [comment]
I mean that a simple renderer can just extrude shapes (pixels into voxels):

![Voxel Mario](http://img06.deviantart.net/f6f1/i/2006/052/3/1/super_mario_bros_3d_by_kritter5x.png)

There are other examples to be sure, but that's what I mean by rendering a game in 2D or 3D by only swapping the renderer. With an isometric game, you really need 3D coordinate space, anyway. The isometric renderer can be easily swapped out with orthographic or perspective renderers...
